monai/config/typing.py

```diff
@@ -0,0 +1,28 @@
+# Copyright 2020 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This file contains type information that should be used consistently throughout
+# the entire MONAI package.
+
+from typing import Hashable, Iterable, TypeVar
+
+# The MonaiDictionaryKeySelection type is used to for defining variables
+# that store a subset of keys to select items from a dictionary.
+# The keys must be hashable, and if a container of keys is specified, the
+# container must be iterable.
+MonaiDictionaryKeySelection = TypeVar("MonaiDictionaryKeySelection", Iterable[Hashable], Hashable)
```

Comment:
maybe it's personal preference but I think this wrapper hurts the code readability. if this is really needed I think we should do this in a 'pyi' file instead of in the codebase. what do you think? cc @Nic-Ma @ericspod

Reply:
@wyli  This comes from my experiences with large C++ code bases.  It helps to assign "conceptual" names to the underlying storage representation.  

The classic example is that you may may store "AgeInYears", "HouseNumber" and "NumberOfSiblings" as 32bit unsigned integers, but conceptually they represent different types of attributes.

Using these concepts also makes maintenance much easier.  In my silly example above, if at some point in the future we need to change "AgeInYears" to contain fractional part of years, we need only change that concept for int32 to float.

Additionally, it provides a glossary of common concepts that help new users to understand how different concepts interact.

That's my 2cents worth. My past experience has been that expressing domain concepts is more important than knowing language types.

Reply:
@Nic-Ma, @ericspod Would you mind providing your review of this PR?  It is the next item that must be resolved and is holding up a large amount of completed work that depends on the decisions made in this PR.

Thanks,
Hans

Reply:
Hi guys, I think maybe `Union[Iterable[Hashable], Hashable]` is more clear for a type hints?
@ericspod , what do you think?

Reply:
I agree that we should have names to concepts like this. In C/C++ we would use typedef to do this so that the concept is represented by a new type name but is also one which can be different depending on environment. We do need to be consistent with using the concept however and record in documentation the rationale and convention behind it. For readability I would strongly suggest much shorted names, people are more willing to use these sorts of things if it isn't onerous to remember and type out. There is no need to prepend "Monai" for example to the names, so `MonaiDictionaryKeySelection` could be `DictKeySelection` instead and still retain the meaning of the name. Putting these names into a source file with a more descriptive name like `type_definitions.py` would help to explicitly identify what is defined, and so the type would be named as `type_definitions.DictKeySelection` which states what we want to say in the name itself.

On a more general practice level, if you're repeating the same concept many times over in your code then it's something that should be named and defined once then reused everywhere. Writing the same few lines of code obviously begs it to be placed in a function with an appropriate name plus documentation, the same set of cooperating methods maybe make sense in a mixin, etc. If we're using the same pattern of typing in many places, we can name that pattern so making it concrete with a new definition makes sense. That nails the concept to the wall and creates something documentable too.

Reply:
@ericspod Thank you for the in-depth explanation.  You have very nicely summarized my convictions regarding the importance of named concepts.

@wyli This should now be ready for merging.

Reply:
thanks, the type is now improved to `DictionaryKeysCollection `, could we have even simpler type names? sorry for being picky, since we'll work with these types everyday, really hope they can be simple and intuitive

Reply:
> Hi guys, I think maybe `Union[Iterable[Hashable], Hashable]` is more clear for a type hints?
> @ericspod , what do you think?

I did review this, and identified that what was really needed was a "Collection[Hashable]".  Once type hint checking was able to determine the values, it was obvious that "Hashable" is not a valid type for some parts of the code.  The "keys" have always been plural, so I think that enforcing that a collection is provided is OK.  Also, some parts of the code need to get "len(keys)" so an iterable is also not sufficient.

Reply:
> thanks, the type is now improved to `DictionaryKeysCollection `, could we have even simpler type names? sorry for being picky, since we'll work with these types everyday, really hope they can be simple and intuitive

@wyli I'll make it whatever you want. How about KeysCollection since "Keys" implies "from Dictionary"

FYI:   I subscribe to the tenant that code is read many many more times than it is written.  The vast number of "clean code", "refactoring", and "software development for distributed teams" publications that I have read state that it is a mistake to make compact names for the purpose of saving keystrokes.  Variable and function names should be focused on reducing time for human comprehension as the primary goal.  Given that almost all code editors provide auto-completion, hinting, and even AI-driven auto-coding, longer variable names help with the most critical task of all: "human comprehension".